`SharedMemory` and `ndarray` serialization

tpietzsch opened the pull request:
This PR adds a class `ShmNDArray` that annotates SharedMemory with dtype and shape.
If NumPy is available `ShmNDArray.ndarray()` provides a `numpy.ndarray` view.

See this [usage example](https://github.com/apposed/appose-java/blob/65ed5cf2227b29605e390b65a8ca3ad55ede238e/src/test/java/org/apposed/appose/NDArrayExamplePython.java) in the corresponding https://github.com/apposed/appose-java/pull/5 PR.

tpietzsch committed:
Add forklift channel for flake8 on aarch64 macs

tpietzsch committed:
Deserialize appose_type "shm" and "ndarray"

ctrueden committed:
Fix linting errors

ctrueden reviewed (commented):
Thank you, this looks great!

I have one question for discussion before we merge this: on the Java side in apposed/appose-java#5, the analogous class is named just `NDArray` rather than `ShmNDArray`. Should we be consistent with the name between the two Appose implementations? Maybe the Java class should be called `ShmNDArray` as well?

ctrueden committed:
Rename ShmNDArray to NDArray

For consistency with appose-java.

ctrueden committed:
Import SharedMemory class directly

ctrueden committed:
Add docstrings to NDArray class

ctrueden committed:
Add type hint for NDArray shape parameter

ctrueden committed:
Swap order of NDArray constructor args

By putting shm last, we can give it a default value of None,
so that the SharedMemory can be created on the fly if not given.

See also apposed/appose-java@434603673eb6354da6b785ef8125a55ebc88ff20

ctrueden committed:
Avoid using `type` as a variable name

Let's not shadow the built-in type function.

ctrueden committed:
Support JSON-encoding of NDArrays

ctrueden committed:
Add tests for SharedMemory and NDArray

ctrueden committed:
Add tests for JSON en/de-code functions

ctrueden committed:
Make encoded JSON as compact as possible

Not likely to make an actual performance difference, but may as well.
This also better matches the JSON-encoding behavior of appose-java.

ctrueden committed:
Make appose.SharedMemory and appose.NDArray work

So that these classes are in the same package (the base one)
as the Java codebase's equivalents.

And so that people don't need to know to import SharedMemory from
multiprocessing.shared_memory, which is kind of a keyboardful.

ctrueden committed:
Add docstring for Python worker

ctrueden committed:
Disable Python worker's resource tracker

To keep things manageable, we never want Python worker processes
cleaning up shared memory blocks. We leave it to the service process
to do that, always.

ctrueden committed:
Remove obsolete TODO comments

ctrueden committed:
Let SharedMemory and NDArray support `with` blocks

To do this, we subclass Python's SharedMemory class,
rather than simply passing it along verbatim anymore.

And we implement the __exit__ method to call the Appose SharedMemory
subclass's dispose() method, which calls either close() or unlink()
depending on whether the unlink_on_dispose flag is set.

Note that this implementation cannot quite align with the Java one,
because in Java, the AutoCloseable interface always calls close().
As such, the close() method must be overridden and taught to sometimes
call unlink(), and sometimes not, depending on the unlinkOnClose flag.
